extensions/amp-sidebar/0.1/amp-sidebar.js

```diff
@@ -0,0 +1,163 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout} from '../../../src/layout';
+import {assert} from '../../../src/asserts';
+import {isExperimentOn} from '../../../src/experiments';
+import {log} from '../../../src/log';
+
+/** @const */
+const EXPERIMENT = 'amp-sidebar';
+
+/** @const */
+const TAG = 'AmpSidebar';
+
+const WHITELIST_ = ['AMP-ACCORDION', 'AMP-FIT-TEXT', 'AMP-IMG'];
+
+class AmpSidebar extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @const @private {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), EXPERIMENT);
+
+    /** @private @const {!Element} */
+    this.documentElement_ = this.getWin().document.documentElement;
+
+    /** @private @const {string} */
+    this.direction_ = this.element.getAttribute('direction');
+
+    /** @private @const {boolean} */
+    this.hasMask_ = false;
+
+    if (this.direction_ != 'left' && this.direction_ != 'right') {
+      this.element.setAttribute('direction', 'left');
+    }
+
+    if (!this.isExperimentOn_) {
+      log.warn(TAG, `Experiment ${EXPERIMENT} disabled`);
+      return;
+    }
+    if (!this.checkWhitelist_()) {
+      return;
+    }
+
+    if (this.documentElement_.classList.contains('amp-sidebar-open')) {
+      // Create the mask if the sidebar is rendered in open mode.
+      this.createMask_();
+    }
+
+    this.documentElement_.addEventListener('keydown', event => {
+      // Close sidebar on ESC.
+      if (event.keyCode == 27) {
+        this.close_();
+      }
+    });
+    this.registerAction('toggle', this.toggle_.bind(this));
+    this.registerAction('open', () => {
+      this.open_();
+    });
+    this.registerAction('close', () => {
+      this.close_();
+    });
+  }
+
+  /**
+   * Toggles the open/close state of the sidebar.
+   * @private
+   */
+  toggle_() {
+    if (this.documentElement_.classList.contains('amp-sidebar-open')) {
+      this.close_();
+    } else {
+      this.open_();
+    }
+  }
+
+  /**
+   * Reveals the sidebar.
+   * @private
+   */
+  open_() {
+    this.mutateElement(() => {
+      this.createMask_();
+      this.documentElement_.classList.add('amp-sidebar-open');
+      this.documentElement_.classList.remove('amp-sidebar-closed');
+    });
+  }
+
+  /**
+   * Hides the sidebar.
+   * @private
+   */
+  close_() {
+    this.mutateElement(() => {
+      this.documentElement_.classList.remove('amp-sidebar-open');
+      this.documentElement_.classList.add('amp-sidebar-closed');
```

Comment:
Main reason to have 2 classes is to not show animation on the initial state. (when sidebar is hidden )